packages/subgraph/src/PaymentsBook.ts

```diff
@@ -0,0 +1,105 @@
+import { Address, BigInt } from '@graphprotocol/graph-ts'
+
+import { buildId } from '../helpers/utils'
+import { loadOrCreateERC20 } from './ERC20'
+import { createFeeMovement } from './Treasury'
+
+import { GuardiansShare, PaymentReceipt, PaymentsBookModule, PaymentPeriod, GuardianShareClaim } from '../types/schema'
+import { PaymentsBook, PaymentReceived, GuardianShareClaimed, GovernorSharePctChanged } from '../types/templates/PaymentsBook/PaymentsBook'
+
+const PAYMENTS_BOOK_MOVEMENT = 'Payment'
+
+export function handlePaymentReceived(event: PaymentReceived): void {
+  updateCurrentPaymentPeriod(event.address, event.block.timestamp, event.params.token)
+
+  const id = buildId(event)
+  const payment = new PaymentReceipt(id)
+  payment.token = loadOrCreateERC20(event.params.token).id
+  payment.period = event.params.periodId.toString()
+  payment.payer = event.params.payer
+  payment.sender = event.params.sender
+  payment.amount = event.params.amount
+  payment.data = event.params.data
+  payment.createdAt = event.block.timestamp
+  payment.save()
+}
+
+export function handleGuardianShareClaimed(event: GuardianShareClaimed): void {
+  updateCurrentPaymentPeriod(event.address, event.block.timestamp, event.params.token)
+  createFeeMovement(PAYMENTS_BOOK_MOVEMENT, event.params.guardian, event.params.amount, event)
+
+  const shareId = buildGuardianShareId(event.params.guardian, event.params.token, event.params.periodId)
+  const shareClaim = new GuardianShareClaim(shareId)
+  shareClaim.guardian = event.params.guardian.toHexString()
+  shareClaim.period = event.params.periodId.toString()
+  shareClaim.token = event.params.token.toHexString()
+  shareClaim.amount = event.params.amount
+  shareClaim.save()
+}
+
+export function handleGovernorSharePctChanged(event: GovernorSharePctChanged): void {
+  const module = loadOrCreatePaymentsBookModule(event.address)
+  module.governorSharePct = BigInt.fromI32(event.params.currentGovernorSharePct)
+  module.save()
+}
+
+export function updateCurrentPaymentPeriod(address: Address, timestamp: BigInt, token: Address | null = null): void {
```

Comment:
Could the `token` be null here?

Reply:
Yes, based on whether there is a token we are updating the payments corresponding to that token or not. There are some events like a heartbeat that doesn't require updating payments related info